src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineRetentionStrategy.java

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.computeengine;
+
+import com.google.common.collect.ImmutableList;
+import hudson.model.Action;
+import hudson.model.Cause;
+import hudson.model.CauseAction;
+import hudson.model.Executor;
+import hudson.model.ExecutorListener;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.slaves.RetentionStrategy;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy;
+
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A strategy that allows:
+ * - setting one shot instances {@link OnceRetentionStrategy}
+ * - in case of preemption of GCP instance to restart preempted tasks
+ */
+public class ComputeEngineRetentionStrategy extends RetentionStrategy<ComputeEngineComputer> implements ExecutorListener {
+    private static final Logger LOGGER = Logger.getLogger(ComputeEngineRetentionStrategy.class.getName());
+
+    private final OnceRetentionStrategy delegate;
+    private final boolean oneShot;
+
+    /**
+     * Creates the retention strategy.
+     *
+     * @param retentionTimeMinutes number of minutes of idleness after which to kill the slave; serves a backup in case the strategy fails to detect the end of a task
+     * @param oneShot              create one shot instance strategy
+     */
+    ComputeEngineRetentionStrategy(int retentionTimeMinutes, boolean oneShot) {
+        this.oneShot = oneShot;
+        delegate = new OnceRetentionStrategy(retentionTimeMinutes);
+    }
+
+    @Override
+    public long check(ComputeEngineComputer c) {
+        return delegate.check(c);
+    }
+
+    @Override
+    public void start(ComputeEngineComputer c) {
+        delegate.start(c);
+    }
+
+    @Override
+    public void taskAccepted(Executor executor, Queue.Task task) {
+        if (oneShot) {
+            delegate.taskAccepted(executor, task);
+        }
+    }
+
+    @Override
+    public void taskCompleted(Executor executor, Queue.Task task, long durationMS) {
+        checkPreempted(executor, task);
+        if (oneShot) {
+            delegate.taskCompleted(executor, task, durationMS);
+        }
+    }
+
+    @Override
+    public void taskCompletedWithProblems(Executor executor, Queue.Task task, long durationMS, Throwable problems) {
+        checkPreempted(executor, task);
+        if (oneShot) {
+            delegate.taskCompletedWithProblems(executor, task, durationMS, problems);
+        }
+    }
+
+    private Queue.Task getBaseTask(Queue.Task task) {
+        Queue.Task parent = task.getOwnerTask();
+        while (task != parent) {
+            task = parent;
+            parent = task.getOwnerTask();
+        }
+        return parent;
+    }
+
+    private void checkPreempted(Executor executor, Queue.Task task) {
```

Comment:
This method seems to encapsulate two separate concerns: 1. Checking if a task was preempted 2. Rescheduling the task if #1 is true. Suggest separating the concerns for improved orthogonality and easier testing.